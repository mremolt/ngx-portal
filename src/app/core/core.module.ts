import { CommonModule } from '@angular/common';
import { InjectionToken, NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { ActionReducerMap, META_REDUCERS, MetaReducer, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';

import { environment } from '../../environments/environment';
import { AppEffects } from '../app.effects';
import { RootState, metaReducers, reducers } from '../reducers';

export const REDUCER_TOKEN = new InjectionToken<ActionReducerMap<RootState>>('Registered Reducers');

export function getReducers() {
  return reducers;
}

export function getMetaReducers(): MetaReducer<RootState>[] {
  return metaReducers;
}

@NgModule({
  imports: [
    CommonModule,
    // ngrx
    StoreModule.forRoot(REDUCER_TOKEN),
    !environment.production ? StoreDevtoolsModule.instrument() : [],
    EffectsModule.forRoot([AppEffects]),
    StoreRouterConnectingModule.forRoot(),
  ],
  declarations: [],
  providers: [
    {
      provide: REDUCER_TOKEN,
      useFactory: getReducers,
    },
    {
      provide: META_REDUCERS,
      useFactory: getMetaReducers,
    },
  ],
})
export class CoreModule {}
