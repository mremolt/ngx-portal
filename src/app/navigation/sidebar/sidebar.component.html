<nav>
  <h2>Navigation</h2>
  <ul>
    <li><a routerLink="/" routerLinkActive="active">Home</a></li>
    <li><a routerLink="/foo" routerLinkActive="active">Two</a></li>
    <li><a routerLink="/bar" routerLinkActive="active">Three</a></li>
    <li><a routerLink="/baz" routerLinkActive="active">Four</a></li>
  </ul>
</nav>
