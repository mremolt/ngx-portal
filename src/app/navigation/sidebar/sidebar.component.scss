nav {
  h2 {
    margin: 0;
    padding: 1em 1em 0.5em 1em;
    // font-size: 1.5em;
  }

  ul {
    padding: 0;
    margin: 0;

    li {
      list-style: none;
      padding: 0;

      .nav-item {
        display: block;
        cursor: pointer;

        padding: 0.5em 1em 0.5em 1em;
        color: #333;
        font-size: 1.3em;
        text-decoration: none;
        text-align: left;
        border: none;

        &.active {
          background-color: #ccc;
        }

        &:hover {
          background-color: #bbb;
        }
      }

      button.nav-item {
        width: 100%;
        background-color: inherit;

        &.active {
          border: 0;
        }
      }
    }
  }

  .error {
    border: 1px solid rgb(255, 50, 50);
    border-radius: 0.2rem;
    background-color: rgb(255, 150, 150);
    color: rgb(255, 20, 20);
    margin: 0.5rem;

    strong {
      margin-right: 0.3rem;
    }
  }
}
