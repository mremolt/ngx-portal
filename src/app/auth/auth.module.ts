import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { AuthEffects } from './reducers/auth.effects';
import { AuthGuard } from './reducers/auth.guard';
import * as fromAuth from './reducers/auth.reducer';

@NgModule({
  imports: [
    CommonModule,
    StoreModule.forFeature('auth', fromAuth.reducer),
    EffectsModule.forFeature([AuthEffects]),
  ],
  declarations: [],
  providers: [AuthGuard],
})
export class AuthModule {}
