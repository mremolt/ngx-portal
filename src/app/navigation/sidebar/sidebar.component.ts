import { HttpErrorResponse } from '@angular/common/http';
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { Store, select } from '@ngrx/store';
import { Observable } from 'rxjs';

import { Authenticate, Logout } from '../../auth/reducers/auth.actions';
import { AuthState, selectAuthenticated, selectError } from '../../auth/reducers/auth.selectors';

@Component({
  selector: 'mr-sidebar',
  templateUrl: './sidebar.component.html',
  styleUrls: ['./sidebar.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SidebarComponent {
  public authenticated$: Observable<boolean>;
  public error$: Observable<HttpErrorResponse | null>;
  public form: FormGroup;

  constructor(private store: Store<AuthState>, fb: FormBuilder) {
    this.authenticated$ = this.store.pipe(select(selectAuthenticated));
    this.error$ = this.store.pipe(select(selectError));

    this.form = fb.group({
      email: ['', [Validators.required, Validators.email]],
      password: ['', [Validators.required]],
    });
  }

  public logout(): void {
    this.store.dispatch(new Logout());
  }

  public login() {
    if (this.form.valid) {
      const { email, password } = this.form.value;
      this.store.dispatch(new Authenticate(email, password));
    }
  }
}
