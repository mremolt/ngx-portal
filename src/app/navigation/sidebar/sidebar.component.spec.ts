import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { Store, StoreModule } from '@ngrx/store';

import { configureTestBedSuite } from '../../../testing';
import { reducer } from '../../auth/reducers/auth.reducer';
import { SidebarComponent } from './sidebar.component';

describe('SidebarComponent', () => {
  configureTestBedSuite();

  let component: SidebarComponent;
  let fixture: ComponentFixture<SidebarComponent>;

  beforeAll(async () => {
    await TestBed.configureTestingModule({
      declarations: [SidebarComponent],
      imports: [ReactiveFormsModule, StoreModule.forRoot<any, any>({ auth: reducer })],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(SidebarComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('Actions', () => {
    let store: Store<any>;
    let dispatchSpy: jest.Mock;

    beforeEach(() => {
      store = TestBed.get(Store);
      dispatchSpy = jest.spyOn(store, 'dispatch').mockImplementation(() => {});
    });

    describe('form', () => {
      it('validates', () => {
        component.form.setValue({ email: 'test@example.com', password: 'pass' });
        expect(component.form.valid).toBeTruthy();

        component.form.setValue({ email: 'test@example.com', password: '' });
        expect(component.form.valid).toBeFalsy();

        component.form.setValue({ email: 'test', password: 'pass' });
        expect(component.form.valid).toBeFalsy();

        component.form.setValue({ email: 'test-example.com', password: 'pass' });
        expect(component.form.valid).toBeFalsy();
      });
    });

    describe('login', () => {
      it('dispatches Authenticate action with valid form data', () => {
        component.form.setValue({ email: 'test@example.com', password: 'secret' });
        component.login();

        expect(dispatchSpy).toHaveBeenCalledWith({
          payload: { email: 'test@example.com', password: 'secret' },
          type: '[Auth] Authenticate',
        });
      });

      it('does not dispatch any action with invalid form data', () => {
        component.form.setValue({ email: 'no-email', password: 'secret' });
        component.login();

        expect(dispatchSpy).not.toHaveBeenCalled();
      });
    });

    describe('logout', () => {
      it('dispatches Logout action', () => {
        component.logout();

        expect(dispatchSpy).toHaveBeenCalledWith({
          type: '[Auth] Logout',
        });
      });
    });
  });
});
