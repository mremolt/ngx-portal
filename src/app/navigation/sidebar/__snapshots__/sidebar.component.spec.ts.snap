// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarComponent should render 1`] = `
<mr-sidebar
  authenticated$={[Function Store]}
  error$={[Function Store]}
  form={[Function FormGroup]}
  store={[Function Store]}
>
  <nav>
    <h2>
      Navigation
    </h2>
    
    <h3>
      Login
    </h3>
    <form
      class="ng-untouched ng-pristine ng-invalid"
      ng-reflect-form="[object Object]"
      novalidate=""
    >
      <div
        class="form-group"
      >
        <label
          for="input-email"
        >
          E-Mail
        </label>
        <input
          class="form-control ng-untouched ng-pristine ng-invalid"
          formcontrolname="email"
          id="input-email"
          ng-reflect-name="email"
          type="email"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="input-password"
        >
          Password
        </label>
        <input
          class="form-control ng-untouched ng-pristine ng-invalid"
          formcontrolname="password"
          id="input-password"
          ng-reflect-name="password"
          type="password"
        />
      </div>
      <div
        class="form-group"
      >
        <button
          type="submit"
        >
          Login
        </button>
      </div>
      
    </form>
    
  </nav>
</mr-sidebar>
`;
